research/picai/prepare_data.py

```diff
@@ -0,0 +1,209 @@
+        (list(map(str, tup[0])), str(tup[1]), lbl) for tup, lbl in zip(paths_for_each_sample, labels)
+    ]
+
+    if splits_path is None:
+        # If splits_path is None, create a singe dataset overview
+        d = {}
+        d["image_paths"], d["label_paths"], d["case_label"] = zip(*scan_annotation_label_list)
+        write_path = os.path.join(write_dir, "train-fold-all.json")
+        with open(write_path, "w") as f:
+            json.dump(d, f)
+    else:
+        # If splits_path is not None, create a dataset overview for each split
+        with open(splits_path, "r") as splits_f:
+            splits = json.load(splits_f)
+
+        for i, split in enumerate(splits):
+            # Add image (scan) and label (annotation) paths that include any subject ids from split
+            # For both train and val
+            train_d, val_d = {}, {}
+            train_list = [
+                tup
+                for tup in scan_annotation_label_list
+                if any([subject_id in str(tup[1]) for subject_id in split["train"]])
+            ]
+            val_list = [
+                tup
+                for tup in scan_annotation_label_list
+                if any([subject_id in str(tup[1]) for subject_id in split["val"]])
+            ]
+            train_d["image_paths"], train_d["label_paths"], train_d["case_label"] = zip(*train_list)
+            val_d["image_paths"], val_d["label_paths"], val_d["case_label"] = zip(*val_list)
+
+            # Create path of json files and write to disk
+            train_write_path = os.path.join(write_dir, f"train-fold-{i}.json")
+            val_write_path = os.path.join(write_dir, f"val-fold-{i}.json")
+
+            with open(train_write_path, "w") as f:
+                json.dump(train_d, f)
+
+            with open(val_write_path, "w") as f:
+                json.dump(val_d, f)
+
+
+def preprare_data(
+    scans_read_dir: Path,
+    annotation_read_dir: Path,
+    scans_write_dir: Path,
+    annotation_write_dir: Path,
+    overview_write_dir: Path,
+    size: Optional[Sequence[int]] = None,
+    physical_size: Optional[Sequence[int]] = None,
+    spacing: Optional[Sequence[int]] = None,
+    scan_extension: str = "mha",
+    annotation_extension: str = ".nii.gz",
+    num_threads: int = 4,
+    splits_path: Optional[Path] = None,
+) -> None:
+    settings = PreprocessingSettings(
+        scans_write_dir,
+        annotation_write_dir,
+        size,
+        physical_size,
+        spacing,
+    )
+    valid_annotation_filenames = [f for f in os.listdir(annotation_read_dir) if f.endswith(annotation_extension)]
+
+    samples = []
+    for annotation_filename in valid_annotation_filenames:
+        # Annotation filename is subject id (ie patient_id study id)
+        # We use it to get the corresponding scan paths
+        annotation_path = Path(os.path.join(annotation_read_dir, annotation_filename))
+        annotation_base_filename = annotation_filename.split(".")[0]
```

Comment:
I think Path objects have a canonical way to drop the extension that will handle edge cases like `john.jewell.png` (which I think is what you're doing here?

Reply:
Cool!